### ISSUE TITLE
Login panic when using Redis HA with HAproxy

### ISSUE THREAD

**Issue by shahbazn** [goharbor/harbor#10460]
Expected behavior and actual behavior:

Sometimes, core service panic during login. Issue could not be reproduced when using single redis. 

Could be related to #9841 but creating a separate issue because of https://github.com/goharbor/harbor/issues/9841#issuecomment-568843852

**Steps to reproduce the problem:**
Please provide the steps to reproduce this problem.

**Versions:**
Please specify the versions of following systems.

- harbor version: v1.10.0
- docker engine version: 18.9.2
- kubelet version: 1.16.3

**Additional context:**

- **Log files:**
```
Harbor:runtime error: invalid memory address or nil pointer dereference
Request Method:    GET
Request URL:    /c/oidc/callback?code=5I-o7wWFxcudwT8q18OR0D6qSzN5dZZEP07EzFU5yIn&state=dlap3hsued7f23qie9oxov833t5a556a
RemoteAddr:    172.25.14.0
Stack
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/harbor/src/vendor/github.com/astaxie/beego/controller.go:579
/harbor/src/core/api/base.go:180
/harbor/src/core/controllers/oidc.go:183
/usr/local/go/src/reflect/value.go:447
/usr/local/go/src/reflect/value.go:308
/harbor/src/vendor/github.com/astaxie/beego/router.go:815
/usr/local/go/src/net/http/server.go:2774
/usr/local/go/src/net/http/server.go:1878
/usr/local/go/src/runtime/asm_amd64.s:1337
```

**Comment by lvisterin** [comment]
I am also seeing this problem using Harbor v1.10.1. It seems to happen randomly: most of the time login works fine so I cannot reliably reproduce it.

We also using Redis HA with HAProxy and OIDC auth.

**Comment by prudhvigodithi** [comment]
Hey, we are using harbor  v1.10.1 integrated with OIDC, using Redis HA with haproxy, we see this error and after a couple of refreshes this works fine, but in OIDC(keycloak) logs we found that type=CODE_TO_TOKEN_ERROR and error as error=invalid_code, so is this issue related with Redis or OIDC setup cause when harbor tested with LDAP using Redis ha with haproxy we did not get this issue, we only see this issue when using OIDC for harbor.
Redis :
https://github.com/helm/charts/tree/master/stable/redis-ha

**Comment by reasonerjt** [comment]
@prudhvigodithi 
see the comment in 
https://github.com/goharbor/harbor/issues/9841

**Comment by prudhvigodithi** [comment]
Hey reasonerjt thanks for reaching me, I have adjusted few token settings from my OIDC application side, still get the following error.

Harbor:runtime error: invalid memory address or nil pointer dereference
Request Method: | GET
-- | --
Request URL: | /c/oidc/callback?state=2zc5pp3oo3ymtuaoluoj1oz5aaoyd2al&session_state=d4988ac6-439b-43ad-9906-7c223d566bc1&code=70af1b61-7290-4abb-90b0-06d929d8b190.d4988ac6-439b-43ad-9906-7c223d566bc1.f7127b73-6b85-413a-9253-d934194af579
RemoteAddr: | 172.21.184.0

Stack
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/harbor/src/vendor/github.com/astaxie/beego/controller.go:624
/harbor/src/core/api/base.go:180
/harbor/src/core/controllers/oidc.go:146
/usr/local/go/src/reflect/value.go:447
/usr/local/go/src/reflect/value.go:308
/harbor/src/vendor/github.com/astaxie/beego/router.go:853
/usr/local/go/src/net/http/server.go:2774
/usr/local/go/src/net/http/server.go:1878
/usr/local/go/src/runtime/asm_amd64.s:1337

**Comment by lvisterin** [comment]
> this behavior is not the same when the Redis is destroyed and bought up newly, works perfectly fine for a day or a week and then I get the above-mentioned error as "Harbor:runtime error: invalid memory address or nil pointer dereference",
> then I have to again delete and recreate the Redis which again works fine for a few days.

I have seen this issue happen right after running `FLUSHALL` against Redis (completely clearing it) so I think what you described may be a coincidence.

Unfortunately clearing Redis does not always cause this error to happen, it is still random.

**Comment by lvisterin** [comment]
Update: Our Harbor web portal login became completely unusable because of this bug suddenly happening 100% of the time instead of randomly. 

`FLUSHALL` fixed the problem in this case, which is in line with what @prudhvigodithi commented.

**Comment by prudhvigodithi** [comment]
so @lvisterin, thanks for mentioning that FLUSH ALL case, probably when I delete and recreated the Redis cluster and restart the harbor it works because of this FLUSH ALL scenario could have happened.!!

**Comment by ninjaslothx** [comment]
I am having this same issue. 
We are using LDAP, but again redis ha on k8s with an haproxy entrypoint.

**Comment by dvianello** [comment]
Seeing the same, using local users against Haproxy-balanced HA redis.

**Comment by ninjaslothx** [comment]
I can confirm in my case moving to a single standalone redis instance fixes the issue.

**Comment by xunchangguo** [comment]
I am also seeing this problem using Harbor  v1.10.1-f3e11715, with Haproxy-balanced HA redis.
```
2020/05/19 01:25:53.692 [D] [server.go:2774]  |  192.168.18.55| 200 |  16.876213ms|   match| GET      /api/projects/2   r:/api/projects/:id([0-9]+)
2020-05-19T01:25:53Z [ERROR] [/common/api/base.go:73]: GET /api/projects/2/scanner failed with error: {"code":401,"message":"Unauthorized"}
2020/05/19 01:25:53.759 [D] [server.go:2774]  |  192.168.18.55| 401 |   2.346124ms|   match| GET      /api/projects/2/scanner   r:/api/projects/:pid([0-9]+)/scanner
2020-05-19T01:25:53Z [ERROR] [/common/api/base.go:73]: GET /api/users/current/permissions?scope=/project/2&relative=true failed with error: {"code":401,"message":"UnAuthorize"}
2020/05/19 01:25:53.763 [D] [server.go:2774]  |  192.168.18.55| 401 |   2.811581ms|   match| GET      /api/users/current/permissions   r:/api/users/:id/permissions
2020/05/19 01:25:53.781 [D] [server.go:2774]  |  192.168.18.55| 200 |  41.984301ms|   match| GET      /api/repositories   r:/api/repositories
2020-05-19T01:25:53Z [ERROR] [/common/api/base.go:73]: GET /api/users/current/permissions?scope=/project/2&relative=true failed with error: {"code":401,"message":"UnAuthorize"}
2020/05/19 01:25:53.817 [D] [server.go:2774]  |  192.168.18.55| 401 |   4.044133ms|   match| GET      /api/users/current/permissions   r:/api/users/:id/permissions
2020/05/19 01:25:53.854 [D] [server.go:2774]  | 172.20.158.141| 200 |   1.613372ms|   match| POST     /service/notifications   r:/service/notifications
2020-05-19T01:25:56Z [ERROR] [/common/api/base.go:73]: GET /api/users/current failed with error: {"code":401,"message":"UnAuthorize"}
2020/05/19 01:25:56.317 [D] [server.go:2774]  |  192.168.18.55| 401 |   4.539637ms|   match| GET      /api/users/current   r:/api/users/:id
2020/05/19 01:25:56.377 [D] [server.go:2774]  |  192.168.18.55| 200 |   4.268769ms|   match| GET      /api/systeminfo   r:/api/systeminfo
2020/05/19 01:25:56.413 [D] [server.go:2774]  | 192.168.10.179| 200 |   1.861023ms|   match| GET      /api/ping   r:/api/ping
2020/05/19 01:25:59.321 [D] [server.go:2774]  | 192.168.10.179| 200 |   2.170405ms|   match| GET      /api/ping   r:/api/ping
2020/05/19 01:26:02.483 [C] [panic.go:522]  the request url is  /c/login
2020/05/19 01:26:02.483 [C] [panic.go:522]  Handler crashed with error runtime error: invalid memory address or nil pointer dereference
2020/05/19 01:26:02.483 [C] [panic.go:522]  /usr/local/go/src/runtime/panic.go:522
2020/05/19 01:26:02.483 [C] [panic.go:522]  /usr/local/go/src/runtime/panic.go:82
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/runtime/signal_unix.go:390
2020/05/19 01:26:02.484 [C] [panic.go:522]  /harbor/src/vendor/github.com/astaxie/beego/controller.go:624
2020/05/19 01:26:02.484 [C] [panic.go:522]  /harbor/src/core/api/base.go:180
2020/05/19 01:26:02.484 [C] [panic.go:522]  /harbor/src/core/controllers/base.go:116
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/reflect/value.go:447
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/reflect/value.go:308
2020/05/19 01:26:02.484 [C] [panic.go:522]  /harbor/src/vendor/github.com/astaxie/beego/router.go:853
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/net/http/server.go:2774
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/net/http/server.go:1878
2020/05/19 01:26:02.484 [C] [panic.go:522]  /usr/local/go/src/runtime/asm_amd64.s:1337
2020/05/19 01:26:02.484 [server.go:3010]  [HTTP] http: superfluous response.WriteHeader call from github.com/astaxie/beego/context.(*Response).WriteHeader (context.go:230)
```

**Comment by xunchangguo** [comment]
> I can confirm in my case moving to a single standalone redis instance fixes the issue.

is there any way to fixes this issue with redis ha?

**Comment by prudhvigodithi** [comment]
Using AWS Redis Engine I did not get this error again, no haproxy behind redis.

**Comment by prudhvigodithi** [comment]
With AWS Redis engine, it survives up to ~2 weeks and later we have that internal Server error issues, where UI we get that pink banner saying server error and backend image push seems to fail, a quick fix for this is reboot the Redis nodes and restart the Harbor core and job service components, but this would cause some downtime and not sure how to predict when this could happen again, any solution for this, please?

**Comment by phiphi282** [comment]
We found out that beego uses a sessionExists function that just returns false in case of an error.
We are running harbor with oidc auth and a ha redis backend.

After fixing this and adding multiple retries as well as idle connection checking to the redis client we haven't seen this problem again. Since we only tested this since yesterday we can't say for sure if the issue is really gone but at least #532 seems to be working for us again.

We try to add our changes into the next major update for beego.
If you are building your own images you can test out these changes now already by adding this line to the go.mod file:
```
replace github.com/astaxie/beego => github.com/phiphi282/beego v1.12.2
```
I will follow up on this ticket in a few weeks to see if this really solved this issue.

**Comment by reasonerjt** [comment]
@phiphi282 any update?

Could you open an issue to beego and write up a PR?

**Comment by phiphi282** [comment]
@reasonerjt
The changes are already merged to beego, but only for their next major version as it was a bit of a bigger change. 
astaxie/beego#4124

We found another bug where the redis client couldn't handle broken connection which resulted in some users being logged out suddenly so we created an additional pr to change the client and make it a bit more production like (e.g. configurable connection retries) astaxie/beego#4137

We didn't had this problem since we changed our harbor to use the custom beego version but I am not sure when harbor will update to beego 2.
